test/behavior/tautological_int_comparison.zig

```diff
@@ -0,0 +1,108 @@
+const std = @import("std");
+const builtin = @import("builtin");
+
+// All conditional expressions in this test have a guaranteed
+// result, so one branch should never be analyzed.
+// https://github.com/ziglang/zig/issues/2146
+test "tautological integral comparisons" {
+    // Comparing runtime value against oversized constant
+    var u: u4 = undefined;
+    var i: i4 = undefined;
```

Comment:
Edge cases for u4 are `-1,0,1,15,16`, edge cases for i4 are `-9,-8,-1,0,1,7,8`.
Also: Might be nice to show what happens, if one swaps the order.

Reply:
I started chopping down the test cases because the test file was getting large. But I'm down to have an exhaustive matrix of cases here if the value to noise ratio is acceptable. (Combining your suggested values with the various operators will make it pretty long.)

> if one swaps the order

Actually had that more of that previously, and you can see remnants in the comparisons with `large_constant` below. But again, happy to add more of that back!

Maybe I can make a helper with some comptime shenanigans to generate the test cases as a way of taming the noise... I'll experiment.

Reply:
@matu3ba Ok, I gave this much more thorough coverage via a helper, and I think it's even more legible too.

- Upon review, I don't think the "Decls instead of literals" or "Comparing against return values" cases were adding value, so I ejected them.
- Swapped comparison order, runtime var, and comptime undefined are tested for all types.
- Note that `0` and `1` are not always valid egde cases the way this test is set up, because they lie inside the bounds of some of the types being tested. The whole test is predicated on either the true or false branch being guaranteed.
- I didn't like usage of the word "tautological" anymore -- it's vestigial from the original proposal -- so I renamed the test file.

Let me know what you think of the revision! If you want to refer to the original: https://github.com/ziglang/zig/compare/master...hryx:zig:comparus-tautologicus-scratch